core/flyout.js

```diff
@@ -833,3 +1024,147 @@ Blockly.Flyout.terminateDrag_ = function() {
   Blockly.Flyout.startBlock_ = null;
   Blockly.Flyout.startFlyout_ = null;
 };
+
+/**
+ * Compute height of flyout.  Position button under each block.
+ * For RTL: Lay out the blocks right-aligned.
+ * @param {!Array<!Blockly.Block>} blocks The blocks to reflow.
+ * @param {boolean} opt_skip_resize Possibly skip resizing, since sometimes a
+ *    resize will be called immediately anyway.
+ */
+Blockly.Flyout.prototype.reflowHorizontal = function(blocks) {
+  this.workspace_.scale = this.targetWorkspace_.scale;
+  var flyoutHeight = 0;
+  var margin = this.CORNER_RADIUS;
+  for (var i = 0, block; block = blocks[i]; i++) {
+    flyoutHeight = Math.max(flyoutHeight, block.getHeightWidth().height);
+  }
+  flyoutHeight *= this.workspace_.scale;
+  flyoutHeight += margin * 1.5 + Blockly.Scrollbar.scrollbarThickness;
+  if (this.height_ != flyoutHeight) {
+    for (i = 0, block; block = blocks[i]; i++) {
+      var blockHW = block.getHeightWidth();
+      if (block.flyoutRect_) {
+        block.flyoutRect_.setAttribute('width', blockHW.width);
+        block.flyoutRect_.setAttribute('height', blockHW.height);
+        // Rectangles behind blocks with output tabs are shifted a bit.
+        var tab = block.outputConnection ? Blockly.BlockSvg.TAB_WIDTH : 0;
+        var blockXY = block.getRelativeToSurfaceXY();
+        block.flyoutRect_.setAttribute('y', blockXY.y);
+        block.flyoutRect_.setAttribute('x',
+            this.RTL ? blockXY.x - blockHW.width + tab : blockXY.x - tab);
+        // For hat blocks we want to shift them down by the hat height
+        // since the y coordinate is the corner, not the top of the hat.
+        var hatOffset =
+            block.startHat_ ? Blockly.BlockSvg.START_HAT_HEIGHT : 0;
+        if (hatOffset) {
+          block.moveBy(0, hatOffset);
+        }
+        block.flyoutRect_.setAttribute('y', blockXY.y);
+      }
+    }
+    // Record the height for .getMetrics_ and .position.
+    this.height_ = flyoutHeight;
+    Blockly.fireUiEvent(window, 'resize');
+  }
+};
+
+/**
+ * Compute width of flyout.  Position button under each block.
+ * For RTL: Lay out the blocks right-aligned.
+ * @param {!Array<!Blockly.Block>} blocks The blocks to reflow.
+ * @param {boolean} opt_skip_resize Possibly skip resizing, since sometimes a
+ *    resize will be called immediately anyway.
+ */
+Blockly.Flyout.prototype.reflowVertical = function(blocks) {
```

Comment:
I know this code isn't new, but maybe you can explain it to me in person?  It'll help me review the reflowHorizontal function better. I'd also like to make sure I understand why it is so complicated (and costly) to calculate the size of the flyout.